index.mjs

```diff
@@ -0,0 +1,165 @@
+} from 'vscode-languageserver/node.js'
+import {TextDocument} from 'vscode-languageserver-textdocument'
+
+/**
+ * Convert a unist point to a language server protocol position.
+ *
+ * @param {import('unist').Point} point
+ * @returns {Position}
+ */
+function unistPointToLSPPosition(point) {
+  return Position.create(point.line - 1, point.column - 1)
+}
+
+/**
+ * Convert a unist position to a language server protocol range.
+ *
+ * If no position is given, a range is returned  which represents the beginning
+ * of the document.
+ *
+ * @param {import('unist').Position?} position
+ * @returns {Range}
+ */
+function unistLocationToLSPRange(position) {
+  if (!position) {
+    return Range.create(0, 0, 0, 0)
+  }
+
+  const start = unistPointToLSPPosition(position.start)
+
+  return Range.create(
+    start,
+    // Fall back to start if the end position contains null values
+    position.end.line && position.end.column
+      ? unistPointToLSPPosition(position.end)
+      : start
+  )
+}
+
+/**
+ * Convert a vfile message to a language server protocol diagnostic.
+ *
+ * @param {import('vfile-message').VFileMessage} message
+ * @param {string|undefined} source
+ * @returns {Diagnostic}
+ */
+function vfileMessageToDiagnostic(message, source) {
+  const diagnostic = Diagnostic.create(
+    unistLocationToLSPRange(message.position),
+    message.reason,
+    message.fatal ? DiagnosticSeverity.Error : DiagnosticSeverity.Warning,
+    message.ruleId
+      ? message.source
+        ? message.source + ':' + message.ruleId
+        : message.ruleId
+      : message.source || undefined,
+    source
+  )
+  if (message.url) {
+    diagnostic.codeDescription = {href: message.url}
+  }
+
+  return diagnostic
+}
+
+/**
+ * Convert language server protocol text document to a vfile.
+ *
+ * @param {TextDocument} document
+ * @returns {import('vfile').VFile}
+ */
+function lspDocumentToVfile(document) {
+  return new VFile({
+    // VFile expects a file path or file URL object, but LSP provides a file URI
+    // as a string.
+    path: new URL(document.uri),
+    value: document.getText()
+  })
+}
+
+/**
+ * @param {import('vscode-languageserver').Connection} connection
+ * @param {import('vscode-languageserver').TextDocuments<TextDocument>} documents
+ * @param {import('unified-engine').Options['plugins']} plugins
+ * @param {string} prefix
+ */
+function initUnifiedLanguageServer(connection, documents, prefix, plugins) {
+  /**
+   * Process various LSP text documents using unified and send back the
+   * resulting messages as diagnostics.
+   *
+   * @param {TextDocument[]} textDocuments
+   */
+  function processDocuments(...textDocuments) {
+    engine(
+      {
+        files: textDocuments.map((document) => lspDocumentToVfile(document)),
+        ignoreName: '.' + prefix + 'ignore',
+        packageField: prefix + 'Config',
+        pluginPrefix: prefix,
+        plugins,
+        processor: unified(),
+        rcName: '.' + prefix + 'rc',
```

Comment:
I like using a single option `prefix` to fill in these values, but this is up for debate.

Reply:
The remark and rehype CLIs currently use a different package field: `remarkConfig` vs `remark`.
These fields are also toggles: they can be omitted to not allow configuration files, not allow configuration from package.json, or not allow ignore files.

Reply:
I think you meant `remarkConfig` and `rehype`. I think it would be best to make this consistent.

Regardless, I think the language server behaviour should be consistent with the CLI, so I’ll use multiple options for this to mirror the that.